components/sup/src/util/pkg.rs

```diff
@@ -63,6 +65,8 @@ where
         &InstallMode::default(),
         // TODO (CM): pass through and enable ignore-local mode
         &LocalPackageUsage::default(),
+        // Install hooks are run when the supervisor loads the package
+        &InstallHookMode::Ignore,
```

Comment:
yes. I probably did not explain this well enough in the argument comment. We actually run the install hooks in the `add_service` of the Supervisor manager, which makes it unnecesary/repetitive to run them here. Now the reason we call them from `add_service` is because this `install` function is only called if the package was never installed. There a few cases where a package may have been installed but the hooks never ran:

1. The package was installed with `--ignore-install-hooks`
2. The rather common Studio case where the package was built in a studio and its bits were simply saved to `pkg_prefix`
3. The rare but possible case where users provision a node pre loaded with their `/hab/pkgs` files and specs. This is actually a case that exists for our integration tests too.
4. The `install` did run but failed perhaps due to some environmental issue. Rather than just invoke the hook when the package is installed, we give the hook "a chance" to run whenever it is loaded which it would do if it had previously failed. Otherwise the hook may never attempt to run again and the package could be unusable.

Reply:
I'm clarifying the arg comment

Reply:
I think I'm going to need a diagram of all this pretty soon 😅 

If the Supervisor installs a service, without running the install hooks, and the service has dependencies that have install hooks, it seems like those would never get run in this `add_service` scenario.

If a hook fails, I think the only way it'd get retried is if the SpecWatcher determines that it has changes. That seems odd, in that responsibility for retries is de facto delegated to a seemingly unrelated subsystem.

If a service's install hook needs some kind of external configuration, the user will have to know that and pre-load an appropriate `user.toml` file (if we continue with supplying such configuration via `user.toml`, which might have complications when dependency install hooks are considered... and then we're back to point #1 above.)

Reply:
So, setting aside anything that a human manually installs where they want to run an install hook, basically the only times we actually _do_ run the install scripts when we "install" a package is when we're making a Docker container, or actually loading a service in the Supervisor, right?

It seems like we may be changing (and in the process, overloading) the meaning of "install" with this new feature. I wonder if we might want to more clearly delineate the two procedures we now have: putting bits on disk, and then execution of any processes to finalize the installation. I'm thinking of how to avoid situations where we're talking about "installing, but not _really_ installing", which could make it difficult to reason about, both from a coding perspective, but also from a documentation and training perspective.

I don't know that I've got a great suggestion for that right now, but this seems like it'd be worth a discussion at some point before we say this feature is complete and no longer experimental.

Reply:
So I'll try to clarify some of this here.

* If a install hook fails in a Supervisor, it should fail to load and therefore be retried anytime the Supervisor attempts to load the service again.
* Yes a user.toml (or `HAB_<svc name>` env vars) would override a package's default config which I think would be expected.
* You are correct bout when we run the hook: `hab pkg install`, `hab svc load` on a new package, and a docker build.
* I do agree that we have an interesting and convoluting distinction evolving between "installed" and "really installed" here.